<a href="https://colab.research.google.com/github/soyeonkm/datasci101/blob/main/DATASCI_101_Breakout_2_a_Activity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Movie night with pandas
## DATASCI 101 · Group investigation

**Your task:** explore a movie dataset, translate requests into pandas operations, and recommend a film using clearly stated criteria.

Work in **groups of four** on Sections 1–6. Write code in the empty cells and explain your findings in the response spaces. Use the hints when needed. Complete **Section 6C individually**. Section 7 is an optional debugging task.

### Collaboration
Discuss and solve the tasks together. **Formal role rotation applies only to the search-request activity in Sections 3B–3E.** The roles and rotation schedule appear directly before those questions. Other sections use ordinary group discussion, with no assigned roles.

### Working in the notebook
Run cells in order with **Shift+Enter**. Double-click a Markdown response cell to edit it. `#` begins a code comment. Use the table names requested below so later tasks work: `movies` is the full dataset and `films` is its movie-only subset.

| Section | Activity | Suggested time |
|---|---|---|
| 1 | Load and understand the dataset | 3 min |
| 2 | Investigate its structure | 5 min |
| 3 | Search requests and role rotation | 15 min |
| 4 | Check missingness and create a variable | 10 min |
| 5 | Compare title formats | 5 min |
| 6 | Recommend, challenge, and reflect | 12 min |
| 7 | Optional debugging | Extra time |

The core route is about 50 minutes with prior exposure to basic pandas. If time is short, your GSI may assign Section 5 as follow-up.

**Student version:** write your own code and discussion responses. Only imports and the dataset URL are supplied.

## 1. Load and understand the dataset
The July 2024 **Summer Movies** dataset contains movie, TV-movie, and video titles with “summer” in their cleaned names, including words such as “midsummer.” Each row describes **one title**, not one review. The file includes titles with at least 10 IMDb votes.

The data are a historical snapshot, not a current streaming catalog. IMDb ratings summarize participating users' opinions. A high rating does not establish age suitability, and vote count does not measure ticket sales.

### 1A. Import the data
Use the setup information below to load the CSV into a DataFrame named **`movies`**, then display its first **four** rows.

**Hints**
- `pd.read_csv(url)` reads a CSV from a URL and returns a table.
- Assignment with `=` stores a result under a variable name.
- `.head(n)` previews the first `n` rows. `display(...)` shows a table in the notebook.
- Use the supplied raw CSV URL. If the download fails, ask your GSI about the connection or a local copy.

In [17]:
# Provided setup information
import pandas as pd
from IPython.display import display

movies_url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/main/data/2024/2024-07-30/summer_movies.csv"


# Load the CSV into movies, then display its first four rows.

# pd.read_csv(movies_url)
table = pd.read_csv(movies_url)
display(table.head(4))


,tconst,title_type,primary_title,original_title,year,runtime_minutes,genres,simple_title,average_rating,num_votes
0,tt0011462,movie,Midsummer Madness,Midsummer Madness,1920.0,60.0,Drama,midsummer madness,7.4,19
1,tt0026714,movie,A Midsummer Night's Dream,A Midsummer Night's Dream,1935.0,133.0,"Comedy,Fantasy,Romance",a midsummer nights dream,6.8,3931
2,tt0033864,movie,The Teachers on Summer Vacation,Magistrarna på sommarlov,1941.0,86.0,Comedy,the teachers on summer vacation,5.5,78
3,tt0037325,movie,Summer Storm,Summer Storm,1944.0,106.0,"Crime,Drama,Film-Noir",summer storm,6.6,688


### 1B. Interpret the columns
| Column | Meaning |
|---|---|
| `tconst` | Unique IMDb title ID |
| `primary_title` | Display title |
| `title_type` | `movie`, `tvMovie`, or `video` |
| `year` | Release year |
| `runtime_minutes` | Length in minutes |
| `average_rating` | IMDb user rating on a 1–10 scale |
| `num_votes` | Number of IMDb ratings |
| `genres` | Up to three comma-separated genre labels |

**Discuss :**
For each of variables displayed below 'primary_title', is it categorical or quantitative?

**Your answers:**

title_type: categorical

year: quantitative

runtime_minutes: quantitative

average rating: quantitative

num_votes: quantitative

genres: categorical

## 2. Investigate the table
### 2A. Describe its structure
Report the number of rows and columns, list the column names, and display the **last two rows**. Write a one-sentence introduction to the dataset that includes the row unit.

**Hints:** `.shape` returns `(rows, columns)` and `.columns` lists names. These attributes have no parentheses. But `.tail(n)` is a method, so it uses parentheses. A preview does not change the size of the dataset.

In [24]:
# Write your code here.
last_two = table.tail(2)
display(last_two)

last_two.shape
last_two.columns

,tconst,title_type,primary_title,original_title,year,runtime_minutes,genres,simple_title,average_rating,num_votes
903,tt9716208,movie,Summer of Mesa,Summer of Mesa,2020.0,76.0,"Drama,Romance",summer of mesa,5.5,114
904,tt9842874,movie,Nothing Left - Just That Summer,Ein wilder Sommer - Die Wachausaga,2018.0,154.0,"Comedy,Drama,History",nothing left just that summer,4.4,15


Index(['tconst', 'title_type', 'primary_title', 'original_title', 'year',
       'runtime_minutes', 'genres', 'simple_title', 'average_rating',
       'num_votes'],
      dtype='object')

**Our dataset introduction:**

(Double-click to edit.)

### 2B. Count categories and release years
1. List the observed `title_type` values and count how many titles belong to each.
2. Count titles by `year` and present the counts in year order.
3. Identify the most common title type in this file.

**Hints:** select one column with `table["column"]`. `.unique()` lists observed values, `.value_counts()` counts their frequencies, and `.sort_index()` orders the labels. `value_counts()` excludes missing values by default.

**Discuss:**
1. Which title type has the most rows? Give its count.
2. Do the year counts include every movie released that year? Answer yes/no and give one reason.

In [35]:
# Write your code here.

type_counts = table["title_type"].value_counts()
print(type_counts, "\n")

year_counts = table["year"].value_counts()
print(year_counts)

title_type
movie      711
tvMovie    136
video       58
Name: count, dtype: int64 

year
2017.0    35
2008.0    33
2022.0    32
2023.0    32
2015.0    32
          ..
1947.0     1
1953.0     1
1920.0     1
1962.0     1
1928.0     1
Name: count, Length: 83, dtype: int64


**Your answers:**

1. Most common type and count: movie, 711
2. All releases included? Reason: no

## 3. Translate search requests into code
### 3A. Prepare the movie-only table
Create **`films`** by keeping rows of `movies` whose `title_type` equals `"movie"`. Make a separate copy of this subset. Display the first five rows with only **`primary_title`, `runtime_minutes`, and `average_rating`**, in that order. Report the total number of film records.

**Hints**
- `==` tests equality; `=` assigns a result.
- `.loc[condition]` keeps rows for which a condition is True.
- Add `.copy()` when making a subset you intend to modify later.
- A list inside brackets selects multiple columns: `table[["column_a", "column_b"]]`.
- `len(table)` counts all rows; `.head()` previews only five.

**Checkpoint:** before the search rounds, make sure `films` exists and every retained `title_type` is `movie`.

In [40]:
# Write your code here.
films = table.loc[table["title_type"] == "movie"].copy()

print(films[["primary_title", "runtime_minutes", "average_rating"]].head())

print(len(films))



                     primary_title  runtime_minutes  average_rating
0                Midsummer Madness             60.0             7.4
1        A Midsummer Night's Dream            133.0             6.8
2  The Teachers on Summer Vacation             86.0             5.5
3                     Summer Storm            106.0             6.6
4                Centennial Summer            102.0             6.1
711


### Role rotation for Sections 3B–3E only
Label yourselves **A, B, C, and D**. Rotate only for these four search questions.

- **Requester:** states exact criteria, output columns, and ranking rules.
- **Coder:** writes and runs the query.
- **Checker:** checks each condition, sorting, and output. States any error and correction.
- **Recorder:** combines the **request, final code, and checker’s comments** in the round's worksheet record. Also records the match count and a brief result.

| Round | Requester | Coder | Checker | Recorder |
|---|---|---|---|---|
| 1 | A | B | C | D |
| 2 | B | C | D | A |
| 3 | C | D | A | B |
| 4 | D | A | B | C |

**For every round:** agree on the request, write and run the code, then check it. The recorder completes the **single Markdown worksheet record below the code cell**. Paste the final code into its Python block; do not write only “see code above.” State which values or code steps the checker inspected, and record any correction. If no correction was needed, write “None.”

Use `films` as the starting table and a separate result name for each round. Aim for 2–3 minutes per round.

**Share after Round 4:** each recorder shares the entire completed record. Collect all four records in one group notebook in Round 1–4 order, then give every member a copy. Also retain the import cell and the cell creating `films`, so the shared code can be rerun. The four records form the group's complete search worksheet.

**After Section 3E, return to ordinary group discussion without assigned roles.**

### 3B. Round 1: one precise condition
**Requester A:** choose a maximum runtime and ask for all films meeting that limit. State whether a film exactly at the limit qualifies. Choose which output columns will make the result easy to check.

**Coder B:** filter the rows, display a short preview, and report the total number of matches.

**Checker C:** verify the longest retained runtime meets the rule. Compare the preview row count with the total match count.

**Recorder D:** complete the worksheet record below with the exact request, final code, result, and checker’s comments.

**Hints:** `<=` includes the boundary; `<` excludes it. Store the comparison as a named condition if helpful, then use `.loc[condition]`. `.max()` can help check the retained runtime values. In this file, missing runtimes do not satisfy a numerical upper-bound comparison.

**Round 1 working code:** agree on the request, run the code below, then complete the single worksheet record immediately after it.

In [ ]:
# Write the code for the agreed Round 1 request.

### Round 1 worksheet record
**Requester:** A (name: ___) · **Coder:** B (name: ___) · **Checker:** C (name: ___) · **Recorder:** D (name: ___)

**1. Exact request**
Starting table, conditions and boundary rules: ___
Output columns, number of rows, ranking and tie rule: ___

**2. Final code**
Paste the complete corrected code here. Include every definition used in this round except the shared `movies` and `films` setup.

```python
# Paste final code here.
```

**3. Result**
Total matches before selecting top rows: ___
Brief result or pasted output: ___

**4. Checker's comments**
Conditions checked and evidence: ___
Sorting / row-selection check (or “not requested”): ___
Error found and correction (or “None”): ___
Final verdict: matches request / needs correction


### 3C. Round 2: two conditions at once
**Rotate roles. Requester B:** choose a maximum runtime and a minimum vote count. State that **both** conditions must hold. Choose the thresholds before looking at the answer.

**Coder C:** find the qualifying films and display title, runtime, and votes. Report the total count.

**Checker D:** check both conditions separately. Report the maximum retained runtime and minimum retained vote count, or state that no rows matched.

**Recorder A:** complete the worksheet record below with the exact request, final code, result, and checker’s comments.

**Hints:** use `&` for “and,” with parentheses around each comparison: `(condition_a) & (condition_b)`. Use `>=` for an inclusive minimum. Avoid the Python word `and` when combining pandas Series. A zero-row result can be correct; report it before changing the request.

**Round 2 working code:** agree on the request, run the code below, then complete the single worksheet record immediately after it.

In [ ]:
# Write the code for the agreed Round 2 request.

### Round 2 worksheet record
**Requester:** B (name: ___) · **Coder:** C (name: ___) · **Checker:** D (name: ___) · **Recorder:** A (name: ___)

**1. Exact request**
Starting table, conditions and boundary rules: ___
Output columns, number of rows, ranking and tie rule: ___

**2. Final code**
Paste the complete corrected code here. Include every definition used in this round except the shared `movies` and `films` setup.

```python
# Paste final code here.
```

**3. Result**
Total matches before selecting top rows: ___
Brief result or pasted output: ___

**4. Checker's comments**
Conditions checked and evidence: ___
Sorting / row-selection check (or “not requested”): ___
Error found and correction (or “None”): ___
Final verdict: matches request / needs correction


### 3D. Round 3: turn “top movies” into a ranking rule
**Rotate roles. Requester C:** choose a minimum vote count, a ranking variable (`average_rating` or `num_votes`), and how many top titles to return. Specify the sort direction and what to do if scores tie.

**Coder D:** filter all films by the vote rule, sort the eligible table, then select the requested number of titles. Include rating and votes in the output.

**Checker A:** verify the vote threshold and ranking. Confirm that the code sorts the full eligible set before taking the first rows.

**Recorder B:** complete the worksheet record below with the exact request, final code, result, and checker’s comments.

**Hints:** `.sort_values("column", ascending=False)` sorts descending. `.head(n)` selects the first `n` rows in their current order. For multiple sort keys, supply a list of column names and a matching list of directions. Decide the tie rule first; the unique `tconst` ID can serve as a final alphabetical tie-breaker.

**Round 3 working code:** agree on the request, run the code below, then complete the single worksheet record immediately after it.

In [ ]:
# Write the code for the agreed Round 3 request.

### Round 3 worksheet record
**Requester:** C (name: ___) · **Coder:** D (name: ___) · **Checker:** A (name: ___) · **Recorder:** B (name: ___)

**1. Exact request**
Starting table, conditions and boundary rules: ___
Output columns, number of rows, ranking and tie rule: ___

**2. Final code**
Paste the complete corrected code here. Include every definition used in this round except the shared `movies` and `films` setup.

```python
# Paste final code here.
```

**3. Result**
Total matches before selecting top rows: ___
Brief result or pasted output: ___

**4. Checker's comments**
Conditions checked and evidence: ___
Sorting / row-selection check (or “not requested”): ___
Error found and correction (or “None”): ___
Final verdict: matches request / needs correction


### 3E. Round 4: compose your own search
**Rotate roles. Requester D:** choose a release-year range and add **one** runtime, vote-count, or rating condition. Request a ranked shortlist with specific columns. State whether each endpoint of the year range qualifies.

**Coder A:** translate the request into code and explain the operations in order.

**Checker B:** inspect the year boundaries, additional condition, ranking, and treatment of missing year or runtime. Verify that the query starts from the intended table.

**Recorder C:** complete the worksheet record below with the exact request, final code, result, and checker’s comments.

**Hints:** a range requires a lower and an upper comparison. Combine all required comparisons with `&`. Check the chosen columns with `.min()` or `.max()` when useful. Empty output does not prove the code is wrong.

**Discuss:** does the query include both year endpoints? Point to the comparison operators.

**Round 4 working code:** agree on the request, run the code below, then complete the single worksheet record immediately after it.

In [ ]:
# Write the code for the agreed Round 4 request.

### Round 4 worksheet record
**Requester:** D (name: ___) · **Coder:** A (name: ___) · **Checker:** B (name: ___) · **Recorder:** C (name: ___)

**1. Exact request**
Starting table, conditions and boundary rules: ___
Output columns, number of rows, ranking and tie rule: ___

**2. Final code**
Paste the complete corrected code here. Include every definition used in this round except the shared `movies` and `films` setup.

```python
# Paste final code here.
```

**3. Result**
Total matches before selecting top rows: ___
Brief result or pasted output: ___

**4. Checker's comments**
Conditions checked and evidence: ___
Sorting / row-selection check (or “not requested”): ___
Error found and correction (or “None”): ___
Final verdict: matches request / needs correction


**Worksheet handoff:** share this whole record, including its code block. Combine Round 1–4 records in order and distribute the complete notebook to all four members. Check that no record is missing its request, code, or checker comments.

## 4. Check data quality and interpret measurements
### 4A. Audit missing runtimes
Using the **full `movies` table**, count missing and observed `runtime_minutes`. Verify that these counts add up to the total number of rows.

**Hints:** `.isna()` marks missing entries True. `.sum()` counts those True values. `.count()` counts observed entries; `len(table)` counts all rows.

**Discuss:**
1. If a missing runtime is misrecorded as 0, would it pass a `<= 120` filter?
2. Will your shortlist exclude unknown runtimes? State your rule.

In [ ]:
# Write your code here.

**Your answers:**

1. Would 0 pass? ___
2. Unknown-runtime rule: ___

### 4B. Express runtime in hours
Create **`runtime_hours`** in `films` by converting minutes to hours. Display the first five titles with both runtime columns, rounding the displayed values to two decimals. Keep the stored values unrounded.

**Hints:** column arithmetic acts row by row. Assign the calculated Series to `table["new_column"]`. Divide by 60 for this conversion. `.round(2)` can be applied to the displayed table without assigning it back.

**Check:** Does dividing a missing runtime by 60 produce zero or a missing value?

In [ ]:
# Write your code here.

**Your answers:**

Result for missing runtime: ___

### 4C. Describe a typical runtime
Calculate the mean and median `runtime_minutes` in **`films`**. Label the units.

**Hints:** `.mean()` and `.median()` summarize a numerical column and skip missing values by default.

**Discuss:**
1. Is the observed mean runtime above or below the median?
2. Which summary is usually more affected by one extremely long film: mean or median?

In [ ]:
# Write your code here.

**Your answers:**

1. Mean above/below median: ___
2. More affected summary: ___

## 5. Compare title formats
### 5A. Summarize ratings by type
Using **`movies`**, calculate median `average_rating` separately for each `title_type`. Also report how many titles belong to each type.

**Hints:** start with the full table, use `.groupby("grouping_column")`, select the numerical column, then apply `.median()`. `.value_counts()` gives category counts. `films` contains only one title type, so it cannot support this comparison.

**Discuss:**
1. Which title type has the highest median (average )rating? Give the value.
2. Does this comparison prove that title format causes higher ratings? Answer yes/no.

In [ ]:
# Write your code here.

**Your answers:**

1. Type and median rating: ___
2. Causation proved? ___

## 6. Make and challenge a movie-night recommendation
### 6A. Produce a shortlist
As a group, agree on a **maximum runtime** and a **minimum vote count**. Start from `films` and produce up to **three candidates**, ranked by rating descending, then votes descending, then `tconst` alphabetically for any remaining ties.

Display **title, runtime, rating, and votes**. Report the number of eligible films **before** taking the top three. Apply the missing-runtime policy you agreed on in Section 4.

**Hints:** choose criteria in words; build the row conditions; select eligible records; sort the complete eligible table; then select the first three. Keep intermediate results under clear names. Define your own output-column list if helpful; there is no pre-created `display_columns` variable.

**Peer check:** ask another group or group member to identify your runtime limit, vote threshold, and tie rule. Add any missing detail before coding.

**Discuss:** name your chosen film and give its runtime and rating.

**Your answers:**

Runtime limit: ___
Minimum votes: ___
Missing-runtime policy: ___
Tie rule: ___
Detail clarified, if any: ___

In [ ]:
# Build the shortlist and report the total eligible count.

**Your answers:**

Chosen film: ___
Runtime (minutes): ___
Rating: ___

### 6B. Change one rule
Change **only one threshold** from your first search. Predict what will happen to the eligible set, then repeat the analysis using the same ranking and tie rules. Preserve both outputs so you can compare them.

**Deliverable:** two shortlist tables, their eligible counts, and a two-sentence comparison. State whether your recommended film changes. If a search has no matches, report that before relaxing the rule.

**Discuss:**
1. What were the eligible counts before and after the change?
2. Did your chosen film change? Name it before and after.

**Hints:** return to `films` for the new search so you can both widen and narrow the criteria. Reusing an already filtered table may prevent previously excluded films from re-entering. A stricter rule cannot add eligible films when all other rules stay fixed.

**Changed threshold and prediction:**

(Double-click to edit.)

In [ ]:
# Run the revised search from films and preserve the first result.

**Your answers:**

1. Eligible counts, before/after: ___
2. Chosen film, before/after: ___

## 7. Optional: diagnose a plausible but incorrect result
A teammate wants the **five highest-rated films** and writes:

```python
films.head(5).sort_values("average_rating")
```

Identify the incorrect operation order and sort direction. Write corrected code, using votes descending and then `tconst` ascending to break rating ties.

**Hints:** inspect the order of operations and the default sort direction. Code that runs without an error can still answer the wrong question.

In [ ]:
# Write your code here.

**Your answers:**

Order error: ___
Direction error: ___
Tie rule used: ___

## Reference and troubleshooting
This reference gives general syntax patterns, not completed task solutions. Replace placeholder names with names from your own work.

| Goal | Pattern |
|---|---|
| Load a table | `pd.read_csv(url)` |
| Dimensions or names | `table.shape`, `table.columns` |
| Preview rows | `table.head(n)`, `table.tail(n)` |
| Select columns | `table[["column_a", "column_b"]]` |
| Select matching rows | `table.loc[condition]` |
| Combine conditions | `(condition_a) & (condition_b)` |
| Count category values | `series.value_counts()` |
| Rank a table | `table.sort_values("column", ascending=False)` |
| Count missing entries | `series.isna().sum()` |
| Count observed entries | `series.count()` |
| Summary within groups | `table.groupby("group")["value"].median()` |

**Common checks**
- `NameError`: complete and run the cell creating the named table. Check spelling.
- `KeyError`: compare your column name with `movies.columns`.
- `SyntaxError`: check quotes, brackets, and parentheses.
- Ambiguous truth value: use `&` or `|` with parentheses around each pandas comparison.
- Unexpected count: check the starting table, inequality direction, and missing-value policy.
- Unexpected ranking: check sort direction and whether sorting precedes row selection.

### Sources and scope
[TidyTuesday Summer Movies, 2024-07-30](https://github.com/rfordatascience/tidytuesday/blob/main/data/2024/2024-07-30/readme.md) provides the dictionary and extraction code. [IMDb non-commercial dataset terms](https://developer.imdb.com/non-commercial-datasets/) apply to the underlying data. All ratings and vote counts describe this historical snapshot.